[클린코드 리액트 2기 이영우] 계산기 구현

zestlee1106 opened the pull request:
안녕하세요, 리뷰어님!  
클린코드 TDD 리액트 2기 이영우 입니다.

- [x] 2개의 숫자에 대해 덧셈이 가능하다.
- [x] 2개의 숫자에 대해 뺄셈이 가능하다.
- [x] 2개의 숫자에 대해 곱셈이 가능하다.
- [x] 2개의 숫자에 대해 나눗셈이 가능하다.
- [x] AC(All Clear)버튼을 누르면 0으로 초기화 한다.
- [x] 숫자는 한번에 최대 3자리 수까지 입력 가능하다.
- [x] 계산 결과를 표현할 때 소수점 이하는 버림한다.
- [x] 연산의 결과값이 `Infinity`일 경우 `오류`라는 문자열을 보여준다.

---

### 궁금한 점

1. Styled Component 로 나누는 것이 꼭 필요했을까?
   1. Vue 로 생각해 봤을 땐, 컴포넌트 내에서 이벤트 핸들러가 포함이 되었다면 좋았을 것 같다.
   2. 그러나, Styled Component 내에 이벤트 핸들러가 포함이 되어도 되는 걸까? 그렇게 만드는 게 맞을까?
2. state 가 너무 많은 것 같다
   1. 구현을 하다 보니 점점 필요해져서 많이 만들게 되었는데, 줄일 순 없는 걸까?
   2. 남발하다 보니 서로 의존도가 높아진 것 같다. 어떻게 의존도를 낮츨 수 있을까?
3. typescript 를 사용해야 할까?
   1. 자주 사용한 것이 js 라서 js 로 만들었으나... PR 들을 확인하니 모두 ts 로 작성했다.
   2. 타입 에러를 잡아주기 때문에 편하지만... 더 빠르게 작성이 가능한 것은 js 이다... 하지만 공부이기 때문에 ts 로 작성하는 것이 맞을까?
4. re-render 가 자주 될 것 같은 느낌
   1. react 는 state 를 바꾸면 re-render 가 되는 것으로 알고 있는데, state 가 너무 많다 보니 리렌더도 너무 잦을 것 같다. 어떻게 줄일 수 있을까?
5. 커밋 컨벤션을 맞추는 게 좋을까?
   1. 회사가 아닌 혼자 작업할 땐 간단하게 올리는 편인데, 나중에 히스토리 파악을 위해 컨벤션을 맞추는 편이 좋을 것 같다.

---

커밋 단위를 모호하게 잡아서 보시기 힘들 것 같네요 ㅠㅠ 다음 과제 땐 피처 단위로 자르겠습니다.  
리액트는 간단한 유지보수만 참여했어서 많이 미숙한 것 같은데, 이번 기회를 통해 열심히 하여 실력을 키워 보고 싶습니다  
고생이 많으십니다. 감사합니다!

zestlee1106 committed:
initial commit

zestlee1106 committed:
digit, operation, modiefier 버튼 클릭 핸들 이벤트 함수 추가

zestlee1106 committed:
사칙연산 기능 추가

- 수정할 게 너무나 많다.......

zestlee1106 committed:
manifest 에러 수정

zestlee1106 committed:
isTyping state 추가

- 계속 0으로 바꿔주면 요구사항에 어긋나는듯해 보여 타이핑중이라는 state를 추가하였다

zestlee1106 committed:
필요없는 파일 삭제 및 리드미 업로드

zestlee1106 committed:
리드미 업데이트

junghyeonsu commented on package.json:
```diff
@@ -0,0 +1,39 @@
+  "dependencies": {
+    "@testing-library/jest-dom": "^5.16.5",
+    "@testing-library/react": "^13.4.0",
+    "@testing-library/user-event": "^13.5.0",
```
@testing-library 라이브러리들이 `dependencies`에 들어가야 할까요?
그럼 `dependencies`와 `devDependencies`의 차이점은 뭘까요?

zestlee1106 replied:
CRA 로 리액트 프로젝트를 생성했더니, 기본적으로 들어갔던 것 같습니다!  
## dependency
- 현재 프로젝트를 실행하는 데(런타임시) 필수적인 라이브러리들이 포함
- `yarn add [패키지 명]` 하면 저절로 dependency 에 들어간다.
- vue 로 치면... vue 나 nuxt 이런 패키지들이 포함된다
## devDependency
- 개발할 땐 필요하지만, 런타임에는 필요하지 않은 패키지이다.
- `yarn add -D [패키지 명]` 하면 devDependency 에 들어간다.
- typescript 나 @types 혹은 testing 패키지 같은, 개발 시에는 필요하나 런타임에 필요하지 않은 패키지들이 들어간다.

요렇게 정리를 해 보았습니다

감사합니다! :)

junghyeonsu commented on src/App.js:
```diff
@@ -0,0 +1,126 @@
+const OPERATION_LIST = ["/", "X", "-", "+"];
+const DIGIT_LIST = [9, 8, 7, 6, 5, 4, 3, 2, 1, 0];
+
+function App() {
```
현재 App 컴포넌트에 많은 양의 코드가 들어가있는 것 같아요.

상태도 많고, 상태를 핸들링하는 함수들도 정말 많네요.
이걸 어떻게 하면 분리할 수 있을까요?
보통은 커스텀 훅 이라는 개념을 사용하는데, 커스텀 훅의 개념을 조금 찾아보시고 적용을 해보시면 좋을 것 같아요.
아마 `useCalculate` 라는 훅이 생길 것 같은데, 이건 어디까지나 저의 의견이랍니다.
혹시 시도를 더 해보시고 안되시면 또 말씀주세요.

zestlee1106 replied:
use 커스텀 훅을 추가하였습니다 (useCalculate)

감사합니다!!

junghyeonsu commented on src/App.js:
```diff
@@ -0,0 +1,126 @@
+  const handleDigitClick = (digit) => {
+    if (String(number).length === 3) {
+      return;
+    }
```
`String(number)` 라는 걸 보고 맨 처음에 이게 뭐지? 하고 생각했어요.
제가 보기에 number 라는 이름의 상태 값은 숫자일 것 같은데 이게 String으로 변환되어서 함수에서 사용되고 있네요. 그럼 number라는 이름의 상태 이름이 아니라 다른 이름이 되어야 하지 않을까요ㅎㅎ?

zestlee1106 replied:
타입이 들어간 이름은 피하는 게 맞을 것 같습니다  
digit 이라는 이름으로 변경하겠습니다 감사합니다!

junghyeonsu commented on src/App.js:
```diff
@@ -0,0 +1,126 @@
+    setNumber(Number(`${number}${digit}`));
+  };
+
+  const handleOperationClick = async (_operation) => {
```
요기서 `async` 필요없을 것 같아요!
그럼 `async`는 언제 필요할까요?

`_operation` 변수 앞에는 왜 `_`를 붙여주셨나요?

zestlee1106 replied:
`async` 를 붙이면 프라미스를 반환하는 것으로 알고 있었습니다!  
비동기로 동작해야 할 경우 추가가 필요합니다  
useEffect 에서 setState  를 테스트 해 보다가 들어갔던 것 같습니다 (삭제했습니다!)

상태의 이름인 operation 과 겹쳐서 _ 를 붙였었는데, 안티 패턴으로 생각되어 제거하였습니다!

감사합니다 :)

junghyeonsu commented on src/App.js:
```diff
@@ -0,0 +1,126 @@
+function App() {
+  const [number, setNumber] = useState(0);
+  const [total, setTotal] = useState([]);
+  const [operation, setOperation] = useState("");
```
저는 operation은 상태로 안만들고 계산기를 짤 수 있는 방법을 생각해볼 것 같아요.
입력 값을 기준으로 `=`을 눌렀을 때 계산을 해주는 방식을 생각해볼 수 있지 않을까요?
상태가 많아지면 리렌더링도 많아지고, 관리포인트도 더 늘어나니까요ㅎㅎ

zestlee1106 replied:
현재 부호를 [+, -, *, /] 와 = 두개로 나눠서 작업하였습니다!  
```jsx
<Operations>
  {OPERATION_LIST.map((operation) => (
    <button
      onClick={() => handleOperationClick(operation)}
      key={operation}
    >
      {operation}
    </button>
  ))}
  <button onClick={handleCalculateClick}>=</button>
</Operations>
```
= 를 눌렀을 때 바로 계산하게끔 했는데요,  
아무리 생각을 해 보아도... 부호를 저장하는 상태가 하나 필요할 것 같아서요 ㅠㅠ  
혹시 어떤 방법이 있을까요??

junghyeonsu commented on src/App.js:
```diff
@@ -0,0 +1,126 @@
+
+function App() {
+  const [number, setNumber] = useState(0);
+  const [total, setTotal] = useState([]);
```
total이라는 이름말고 조금 더 자세한 이름이 있지 않을까요?

zestlee1106 replied:
digits 라는 이름으로 변경하였습니다. 감사합니다! :)

junghyeonsu commented on src/App.js:
```diff
@@ -0,0 +1,126 @@
+    if (!isFinite(result)) return "오류";
+
+    return result;
+  };
```
해당 함수는 두 가지 역할을 하고 있는 것 같아요!

1. 계산
2.검증

두 가지를 분리를 하자면 어떻게 할 수 있을까요?
계산을 하는 함수와 검증을 하는 함수를 따로 만들 수 있을 것 같네요!

그리고 "/", "X", "-"와 같은 문자열들은 지금 하드 코딩으로 되어있는데
해당 문자들을 상수로 관리를 할 수 있으면 휴먼 에러(우리들이 실수하는 것)도 방지할 수 있을 것 같네요!

zestlee1106 replied:
계산, 검증 두 가지를 분리하기 위해서

**ASIS**
```javascript
  const getCalculatedNumber = (numberList, operation) => {
    let result = 0;
    switch (operation) {
      case "/":
        result = Math.floor(numberList[0] / numberList[1]);
        break;
      case "X":
        result = numberList[0] * numberList[1];
        break;
      case "-":
        result = numberList[0] - numberList[1];
        break;
      case "+":
        result = numberList[0] + numberList[1];
        break;
      default:
        result = 1;
    }

    if (!isFinite(result)) return "오류";

    return result;
  };
```
**TOBE**
```javascript
  const calculate = (digits, operation) => {
    let result = 0;

    switch (operation) {
      case OPERATION.DIVIDE:
        result = Math.floor(digits[0] / digits[1]);
        break;
      case OPERATION.MULTIPLY:
        result = digits[0] * digits[1];
        break;
      case OPERATION.MINUS:
        result = digits[0] - digits[1];
        break;
      case OPERATION.PLUS:
        result = digits[0] + digits[1];
        break;
      default:
        result = 1;
    }

    return result;
  };

  const checkResultIsFinite = (result) => {
    if (!isFinite(result)) return "오류";

    return result;
  };
```

위처럼 수정해 보았습니다. 감사합니다!  
(상수도 추가하였습니다!)

junghyeonsu reviewed (commented):
안녕하세요 영우님!

제가 코멘트로 조금 남겨봤어요ㅎㅎ
현재 프로젝트가 App 컴포넌트에 모든 상태, 모든 로직들이 들어가있어서 보기가 조금 힘들 수도 있고, 관리도 힘들 수도 있을 것 같아요. 그래서 App 컴포넌트는 정말 해당 프로젝트에 진입점으로 두시고, 파일을 조금 더 분리해서 관리를 해주시면 좋을 것 같아요! 그리고 아래에는 영우님이 적어주신 질문들 제가 한 번 답변드려볼게요.

### Styled Component 로 나누는 것이 꼭 필요했을까?

현재 Components 폴더안에 넣어주신 컴포넌트들이 영우님이 말씀해주신 Styled Component일까요?
사실 컴포넌트는 "재활용이 가능한 UI" 라고 저는 생각을 하는데, 그래서 해당 UI에 이벤트 핸들러가 필요하다면 저는 들어가도 된다고 생각을 해요. 지금 영우님께서 컴포넌트로 작성해주신 것들은 "왜 컴포넌트에 넣었고", 현재 계산기 미션에서 꼭 필요한 분리였나?를 생각해주시면 좋을 것 같아요. 만약에 영우님께서 추후의 프로젝트의 발전가능성을 생각하셔서 그렇게 분리하셨다면 그것도 저는 좋다고 생각합니다.

### state 가 너무 많은 것 같다

저도 이 느낌을 받긴 했었어요! 어떻게 하면 상태를 줄일 수 있을까요?
우선 "이 상태가 꼭 필요한 상태인가?"를 생각해보시면 좋을 것 같아요.
그리고 state가 많아질 때는 reducer 라는 개념을 도입해서 상태를 관리하면 조금 더 깔끔하게 관리할 수 있어요. 현재는 하나의 함수에 여러 상태들이 바뀌는 형태잖아요? reducer를 사용하게 되면 함수에 Action을 넘겨서 Action으로 소통을 할 수 있어요. 예를 들면 아래와 같은 느낌인거죠.

```js
// AS-IS
  const handleModifierClick = () => {
    setNumber(0);
    setOperation("");
    setTotal([]);
    setIsTyping(false);
  };

// TO-BE
// 예를 들면 이런식이라는 거에요! 정답이 아니에용
  const handleModifierClick = () => {
    dispatch({ action: "Reset" });
  };
```

이렇게 하면 무엇이 좋을까요?

위에 있는 함수를 읽으면 처음 이 프로젝트를 보는 사람은 이런 식으로 읽겠죠.
"handleModifierClick? 아 Modifier를 클릭했을 때 발생하는거구나.. 내부동작을 볼까? 음... setNumber(0)?.. 0으로 바꿔주고, 오퍼레이션은 ""을 넣어주네? 왜지? total은 빈배열... 음"
이렇게 읽게 될 것 같네요. (실제로 제가 이렇게 읽었어요ㅎ..) 우선 이 코드를 처음보는 사람은 `handleModifierClick`가 내부적으로 크게 어떻게 동작하는지 크게 관심이 없을 수도 있어요.
만약 제가 적은 함수처럼 넣게된다면 무엇이 좋을까요? 이 코드를 처음보는 사람은 이렇게 읽을 수 있을 것 같네요. "handleModifierClick? 아 reset을 하는구나 ok" 정말 짧게 끝나죠?

[벨로퍼트님의 useReducer 블로그 글](https://react.vlpt.us/basic/20-useReducer.html)을 한 번 읽어보시고 찬찬히 적용해보세요!

### typescript 를 사용해야 할까?

네. 이건 과감하게 말씀드릴 수 있겠네요. 사용을 해야할 것 같아요.
왜일까요? 타입스크립트는 뭐가 좋길래 사람들이 그렇게 많이 사용하는 걸까요.
"타입 에러를 잡아주기 때문에 편하다" 이게 핵심이에요.

자바스크립트로 코드를 작성하시면서 중간중간에 Number, String과 같이 강제로 형변환을 시켜주신 것 같은데 이렇게 작성하시고 저장하시고 잘 되는지 계속 확인을 해야했겠죠? 타입스크립트를 쓰면 이런 부분들이 개발단계(컴파일단계)에서 에러를 잡아주기 때문에 그런 걱정을 하지 않아도 되는거에요.
지금은 작은 프로젝트라서 자바스크립트로 작성하시는게 편할 수 있지만, 나중에 프로젝트가 점점 더 커진다면 컴파일 단계 (개발단계)에서 에러를 잡아주는 것이 정말 편하답니다..! 꼭 적용해보세요.

타입스크립트가 정말 어려운 벽 처럼 느껴지시겠지만 사실 지금 당장도 도입할 수 있어요!
npm install --dev typescript로 설치를 하신다음 tsconfig.json 설정만 조금만 만져주면 바로 시작하실 수 있답니다. 그리고 모든 파일을 한꺼번에 적용을 해야하는 것이 아니라 프로젝트 점진적으로 적용도 할 수 있어요. 그리고 javascript를 허용하게 해주는 tsconfig.json 옵션도 있구요, 아직 타입을 잘 모를 때 그냥 `any`로 전부 박아두고 시작하시면 조금 더 쉽게 시작할 수 있을거에요. 그리고 any를 쓰면 빨간줄이 쳐질텐데 any를 허용해주는 tsconfig.json 옵션도 있으니 tsconfig.json 설정들을 잘 찾아보세요!

### re-render 가 자주 될 것 같은 느낌

이 질문은 2번을 참고해주세요.
그리고 리렌더링 관련해서는 어떤식으로 줄일 수 있을까요?

대표적으로는 메모이제이션 기법이 있겠네요. useMemo, useCallback과 같은 친구들을 이용한 메모이제이션 기법이요! 사실 정~~말 많은 변경을 다루는 프로젝트가 아니면 메모이제이션은 오히려 오버엔지니어링이 될 수도 있어요. 왜냐하면 메모이제이션이라는 것도 사실 어떤 하나의 함수가 실행이 되는 건데, 만약 항상 바뀌는 값을 메모이제이션 한다면 어차피 계속 바뀔 값인데 메모이제이션 함수를 한 번 더 계속 실행을 시키는거니까, 최적화를 한다고 했지만 성능은 더 떨어질수도 있어요. 그래서 항상 잘 살펴보시고 사용해야해요.

### 커밋 컨벤션을 맞추는 게 좋을까?

저도 현업에서는 혼자 하는 프로젝트는 커밋 컨벤션을 쎄게 맞추지는 않아요.
하지만 일관성을 지켜주시면 좋아요. 추후에 해당 커밋을 찾아서 돌아가야하는 경우도 있을수도 있어요.
그럴 때 커밋을 너무 대충적어놨다면 정말 찾기가 힘들겠죠? 그래서 커밋은 최대한 그 내용을 잘 담아주는게 좋아요. 

컨벤션을 맞춘다는 게 무슨 의미일까요? 앞에 chore, feat, refactor... 등등과 같은 단어들을 대표적으로 붙이는데 이게 왜 필요한 걸까요?

사실 전부 커뮤니케이션과 관련이 있다고 저는 생각해요. 커밋 메세지 맨 앞에 chore 라는 문구를 보면 다른 사람이 보기에는 "아, 이 커밋은 간단한 수정사항인가?" 하고 바로 생각을 할 수 있는거죠. 그들만의 약속을 만들어놓는거에요. 그럼 그 문맥안에서 불필요한 의사소통이 줄어들고, 그럼 효율성이 조금이라도 더 올라가겠죠.

영우님이 혼자서 개발하시는 프로젝트에 커밋메세지를 뭘써야하나 너무 많은 생각이 들어가서 효율성이 떨어진다면 커밋 컨벤션을 정말 딱딱 다 맞출 필요는 없어요. 개발에 정답은 절대 없으니까요. 하지만 추후에 협업을 하시게 된다면 이런 부분들은 언젠가 필요하시게 될테니 미리 해두시면 나쁘지 않을 것 없다고 저는 생각해요ㅎㅎ

junghyeonsu reviewed (changes requested):
.

zestlee1106 committed:
[chore] 패키지 dependency 수정

- cra 로 리액트 플젝을 생성하며 저절로 들어간 test 관련 라이브러리들을 삭제하였다
- 더불어 devDependency 와, dependency 에 대해서 알아 보고자 한다.

zestlee1106 committed:
[Refactor] 변수명 수정

- 타입스러운 number 대신 digit 으로 변수명을 수정하였다

zestlee1106 committed:
[Refactor] 필요없는 Promise 삭제

zestlee1106 committed:
[Refactor] private 한 이름으로 만들었던 변수명 수정

zestlee1106 committed:
[Refactor] total > digitList 변수명 수정

zestlee1106 committed:
[Refactor] 변수명을 수정하였다 (자료구조 안 들어가게)

zestlee1106 committed:
[Refactor] 부호들을 상수로 빼냄

zestlee1106 committed:
[fix] 에러 수정

zestlee1106 committed:
[Refactor] 함수를 나눔

zestlee1106 committed:
[Refactor] state 과부하를 줄였다. (use 커스텀 훅 추가)

junghyeonsu commented on README.md:
```diff
@@ -1,60 +1,57 @@
+### 그럼, 구분을 안 해도 괜찮지 않을까?
 
-This project is [MIT](https://github.com/next-step/react-calculator/blob/master/LICENSE) licensed.
+- 개발할 때 가시성을 위해 넣어 주는 것이 좋을 것 같다. (다른 개발자들을 위해...)
```
제가 적으려고 했었는데 영우님이 적으셨군요!

[testing-library 공식 문서](https://testing-library.com/docs/dom-testing-library/install/)에 보니까 우선 `--dev` 옵션줘서 설치를 하도록 하게끔 되어있네요.

또한 다른 개발자분들이랑 협업할 때도 확실하게 이게 런타임 때 포함되는 라이브러리인지 아닌지 명확하게 하는 목적도 있을 것 같아요!

junghyeonsu commented on src/App.js:
```diff
@@ -0,0 +1,62 @@
+          <Total>{digit}</Total>
+          <Digits>
+            {DIGIT_LIST.map((digit) => (
+              <button onClick={() => handleDigitClick(digit)} key={digit}>
```
이렇게 익명 함수를 넘기게 되면 해당 버튼들은 무조건 다시 렌더링이 되게 돼요.
리액트는 props가 변했는지 변하지 않았는지 비교를 하고 다시 렌더링을 진행하게 되는데
익명함수는 항상 새로운 메모리 값을 반환하기 때문에 무조건 렌더링이 일어나게 돼요.

이걸 해결하려면 우선 함수를 익명으로 생성하지 않고, `onClick` 이벤트에 함수 자체를 넘기셔야 해요.
그리고 그 함수는 useCallback과 같은 메모이제이션 도구를 통해서 감싸졌을 때 리액트에서 비교를 하고 그걸 토대로 다시 렌더링 할 지 말 지를 결정한답니다.

[컴포넌트 제대로 만들기](https://hyunseob.github.io/2019/06/02/react-component-the-right-way/) 블로그 글 한 번 읽어보시면 좋을 것 같아요!

junghyeonsu commented on src/App.js:
```diff
@@ -0,0 +1,62 @@
+            ))}
+          </Digits>
+          <Modifiers>
+            <button onClick={handleModifierClick}>AC</button>
```
위의 코멘트 처럼 요기서는 함수를 잘 넘기셨네요!
일관성 측면에서도 그렇고 성능 측면에서도 익명함수는 넘기지 않는 게 좋을 것 같아요!

junghyeonsu commented on src/App.js:
```diff
@@ -0,0 +1,62 @@
+          <Operations>
+            {OPERATION_LIST.map((operation) => (
+              <button
+                onClick={() => handleOperationClick(operation)}
```
익명 함수 코멘트 참고해주세요!

junghyeonsu commented on src/components/Modifiers.js:
```diff
@@ -0,0 +1,17 @@
+`;
+
+function Modifiers({ children }) {
+  return <StyledModifiers className="subgrid">{children}</StyledModifiers>;
```
`className`은 왜 붙여주신걸까요!?

`styled-components`를 사용하는 이유와 조금 관련이 있을 것 같은데요, 사용자가 직접 클래스를 고민하고 작성하기가 어려워서 css-in-js 기술을 사용하면 클래스 이름을 고민하지 않아도 되고, 스타일링도 할 수 있어요. 이런 이유가  `styled-components`를 사용하는 이유중에 "하나" 일 것 같네요!

혹시 다른 이유가 있으셨다면 알려주세요~

junghyeonsu commented on src/components/Operations.js:
```diff
@@ -0,0 +1,15 @@
+import styled from "styled-components";
+
+const StyledOperations = styled.div`
+  grid-area: oper;
```
이런 이름도 조금 더 길게 다른 분들이 이해하기 쉽게 적어주시면 좋을 것 같아요!

junghyeonsu commented on src/components/Total.js:
```diff
@@ -0,0 +1,19 @@
+
+function Total({ children }) {
+  return <StyledTotal>{children}</StyledTotal>;
+}
```
이건 정말 제 생각인데요! 컴포넌트 파일에 작성되는 순서에 관한 얘기에요.

### AS-IS
- 스타일드 컴포넌트
- 컴포넌트

로 되어있는데 이것을

### TO-BE
- 컴포넌트
- 스타일드 컴포넌트

로 작성하는 건 어떨까요?
이건 전적으로 제 취향이기도 한데요, 제 생각은 컴포넌트 파일에 UI와 관련된 코드를 먼저 보여주는 게 컴포넌트 파일의 정체성에 맞다고 생각했어요. 스타일은 해당 컴포넌트의 핵심이 아니라서 컴포넌트 밑에 적는 편이구요.
핵심 로직이 먼저 나오는 게 좋다고 생각해요ㅎㅎ 혹시 생각이 다르시다면 제 생각을 무조건 적용하시지 않으셔도 돼요~

junghyeonsu commented on src/hooks/useCalculate.js:
```diff
@@ -0,0 +1,98 @@
+  MULTIPLY: "X",
+  MINUS: "-",
+  PLUS: "+",
+};
```
OPERATION 상수가 지금 App.js에도 존재하고, App.js에서의 UI에도 그려지는데 사용이 되는 것 같은데
해당 코드를 하나의 파일에서 관리하고 나머지에는 import해서 사용하는 방법은 어떨까요?

상수화를 하는 이유는
1. 변경에 유리 (만약 Divide가 "/"가 아니라 '%"로 바뀐다고 하면 상수화를 잘 해놓고 나머지에서 해당 상수를 import해서 사용한다고 치면 상수의 값만 변경하면 되지만, 이렇게 하드 코딩을 해놓은 곳이 많다면 변경이 어려울 수도 있고 실수로 놓치는 부분도 많을거에요)
2. 휴면 에러 방지 ( 저희가 타이핑 실수하는 것을 방지할 수 있어요 )

junghyeonsu commented on src/hooks/useCalculate.js:
```diff
@@ -0,0 +1,98 @@
+};
+
+function useCalculate() {
+  const [digit, setDigit] = useState(0);
+  const [digits, setDigits] = useState([]);
+  const [operation, setOperation] = useState("");
+  const [isTyping, setIsTyping] = useState(false);
```
다음 미션 때 시간이 난다면 꼭 reducer의 개념을 적용해보세요!

junghyeonsu commented on src/hooks/useCalculate.js:
```diff
@@ -0,0 +1,98 @@
+  const [isTyping, setIsTyping] = useState(false);
+
+  const handleDigitClick = (clickedDigit) => {
+    if (String(digit).length === 3) {
```
"3" 이라는 숫자는 제가 이해하기가 어렵네요!
이걸 매직 스트링이라고 하는데요, 이런 매직 스트링은 이름을 부여해서 보통은 상수로 관리해요.
추후 미션에서는 이런 매직스트링을 최대한 줄여보시면 좋을 것 같아요!

junghyeonsu commented on src/hooks/useCalculate.js:
```diff
@@ -0,0 +1,98 @@
+  const [operation, setOperation] = useState("");
+  const [isTyping, setIsTyping] = useState(false);
+
+  const handleDigitClick = (clickedDigit) => {
```
사실 `handleDigitClick` 이라는 함수는 직관적이지 않은 함수 이름이에요.
조금 더 직관적으로 "함수가 하는 일"에 초점을 맞춰서 함수 이름을 짓게되면 더 좋을 것 같아요!

junghyeonsu commented on src/hooks/useCalculate.js:
```diff
@@ -0,0 +1,98 @@
+    setDigit(Number(`${digit}${clickedDigit}`));
+  };
+
+  const handleOperationClick = (operation) => {
```
추후에는 함수 이름을 신경써보시면 좋을 것 같아요~

junghyeonsu commented on src/hooks/useCalculate.js:
```diff
@@ -0,0 +1,98 @@
+    setIsTyping(false);
+  };
+
+  const handleModifierClick = () => {
```
추후에는 함수 이름을 신경써보시면 좋을 것 같아요~

junghyeonsu commented on src/hooks/useCalculate.js:
```diff
@@ -0,0 +1,98 @@
+    return result;
+  };
+
+  const checkResultIsFinite = (result) => {
```
해당 함수 이름은 잘 지으신 것 같아요 👍

junghyeonsu reviewed (approved):
영우님! 열심히 구현해주셨군요!
이번 PR은 우선 merge를 하고 다음 미션에서 제 피드백들을 반영해보시면 될 것 같아요!

### 코드 코멘트 다는 방법

영우님 아마 이번에 코멘트를 `Add single comment`로 하셔서 다신 것 같아요!
아마 이렇게 하면 리뷰어에게 코멘트를 달 때 마다 알림이 가기 때문에 조금 노이지 할 수도 있어요.
그래서 자기 코드에 리뷰를 다는 방법을 간단하게 알려드리자면

1. `Files changed`에 들어온다.

<img width="873" alt="스크린샷 2023-02-12 오후 12 13 11" src="https://user-images.githubusercontent.com/54893898/218290888-3ba1411c-0ff6-4b9f-8cc1-3928ec6a197a.png">

2. 코멘트를 달고 싶은 코드에 드래그를 해서 코멘트를 입력하고 `start review` 버튼을 누른다.
(아마 사진에는 제가 이미 `start review` 버튼을 클릭해서, `Add review comment` 버튼으로 대체되어 있고 맨 처음에는 `start review`일거에요.

<img width="733" alt="스크린샷 2023-02-12 오후 12 13 46" src="https://user-images.githubusercontent.com/54893898/218290910-255c4215-44ae-486f-a40f-f60ef0498ef2.png">

`start review` 버튼을 눌러도 리뷰어에게는 아무런 알림이 가지않아요. 그리고 `Add review comment`를 해도 아무런 알림이 가지 않을거에요.

3. 셀프 리뷰가 끝나면 오른쪽 위의 `Finish your review` 버튼을 클릭해서 `Comment`를 체크하고 `Submit review`를 클릭한다.

셀프리뷰는 "승인" 또는 "변경 요청"이 아니고 단순 코멘트라서 코멘트에 체크해주시고, `Submit review` 버튼을 누르기 전에는 알림이 가지 않아요. 그래서 `Submit review`를 누르면 리뷰어에게 단 한 개의 알림이 가고, 코멘트를 달 수 있는 거죠.

<img width="684" alt="스크린샷 2023-02-12 오후 12 15 48" src="https://user-images.githubusercontent.com/54893898/218290968-f62c2144-d3f6-43da-a2d2-163f36f9d4df.png">

### 다음 미션 때는 꼭 타입스크립트를..!

다음 미션 때는 타입스크립트로 아예 프로젝트를 시작해보셨음 좋겠어요.
맨 처음 타입스크립트를 도입할 때는 그 효용이 느껴지지 않지만 점점 사용하다보면
타입스크립트 없이는 불편한 느낌을 받을 수 있을거에요.
그런 느낌을 받으려면 우선은 시작을 해봐야 하니까요ㅎㅎ

zestlee1106 committed:
[Refactor] 익명 함수를 넘기지 않도록 수정

- 이벤트 핸들러로 익명 함수를 넘기게 되면 리 렌더가 항상 발생한다

zestlee1106 commented on src/components/Modifiers.js:
```diff
@@ -0,0 +1,17 @@
+`;
+
+function Modifiers({ children }) {
+  return <StyledModifiers className="subgrid">{children}</StyledModifiers>;
```
넵, .subgrid 를 `src/style/GlobalStyle.js` 스타일에서 정의하고 있기 때문에 className 을 지정해 줬습니다!

컴포넌트들에서 해당 클래스를 사용하고 있어서 코드가 중복될 것 같다는 생각에 따로 뺐습니다. (같은 스타일이기 때문에)

혹시 좋은 방법이 있을까요??

junghyeonsu replied:
기존에 제공된 코드군요! (제가 헷갈렸네요.)

사실 제가 코멘트를 드린 이유는 보통 단순 css로 스타일링을 하면 id, class와 같은 선택자를 통해서 스타일링을 하게 되는데
이렇게 했을 때 장점은 css-in-js보다 빠른 성능, css 파일로 확실히 관심사 분리가 된다. 등등이 있을 것이고
단점은 선택자를 정하는데 조금 고민이 들어간다, 중복과 재사용성이 조금 떨어진다, 앱이 커질수록 유지보수가 어려워진다 등이 있을텐데요.
그래서 저는 css-in-js를 사용하신 이유에 선택자 이름을 결정하는 시간들을 줄이기 위함도 있다고 생각해서 className이 왜 들어있지? 왜 같은 생각이었어요ㅎㅎ

만약에 저였으면 css-in-js를 이용해서 SubGrid라는 컴포넌트를 만들어서 제공하거나 했을 것 같네요!
네이티브 css와 css-in-js를 혼용해서 사용할 수도 있지만 너무 남발하면 추후에 유지보수성이 떨어지고 개발자들이 따라가야 하는 정보들이 더 늘어날테니까요.

zestlee1106 committed:
[Refactor] css 명 수정

- 다른 사람들이 알아보기 쉽게 바꾸었다

zestlee1106 committed:
[Refactor] 컴포넌트 - 스타일드 컴포넌트 순서 변경

- 컴포넌트가 먼저 바뀌게끔 수정하였다
- 스타일보다는 컴포넌트의 핵심 로직이 먼저 보이는 게 더 깔끔하다

zestlee1106 committed:
[Refactor] operation 관련 상수화
